2048.py

```diff
@@ -0,0 +1,275 @@
+        while not_sorted_move_right_and_down(column):
+            right_and_down_sort(column)
+    return board
+
+
+def right_or_down_sum(array):
+    for index in range(len(array) - 1, 0, -1):
+        if array[index].value == array[index - 1].value:
+            array[index].value += array[index - 1].value
+            array[index - 1].value = 0
+
+
+def sum_right(board):
+    for row in board:
+        right_or_down_sum(row)
+    return board
+
+
+def sum_down(board):
+    for column in get_columns(board):
+        right_or_down_sum(column)
+    return board
+
+
+def left_or_up_sum(array):
+    for index in range(0, len(array) - 1):
+        if array[index].value == array[index + 1].value:
+            array[index].value += array[index + 1].value
+            array[index + 1].value = 0
+
+
+def sum_left(board):
+    for row in board:
+        left_or_up_sum(row)
+    return board
+
+
+def sum_up(board):
+    for column in get_columns(board):
+        left_or_up_sum(column)
+    return board
+
+
+def get_fields_with_zero_value(board):
+    fields_with_zero_value = []
+    for row in board:
+        for field in row:
+            if field.value == 0:
+                fields_with_zero_value.append(field)
+    return fields_with_zero_value
+
+def place_next_element_on_board(board):
+    if len(get_fields_with_zero_value(board)) > 0:
+        next_field = random.choice(get_fields_with_zero_value(board))
+        next_field.value = 2
+    else:
+        print("No more space on the board. You lost!")
+
+
+def check_if_similar_neighboring(board):
+    for row in board:
+        for index in range(0, len(row)-1):
+            if row[index].value == row[index + 1].value:
+                return True
+    for column in get_columns(board):
+        for index in range(0, len(column)-1):
+            if column[index].value == column[index + 1].value:
+                return True
+    return False
+
+
+
+def game_over(board):
+    if len(get_fields_with_zero_value(board)) > 0:
+        return False
+
+
+def check_if_won(board):
+    winning_number = 16
+    for row in board:
+        for field in row:
+            if field.value == winning_number:
+                return True
+    return False
+
+
+def get_temporary_board_values(board):
+    temporary_board_values = []
+    for row in board:
+        for field in row:
+            temporary_board_values.append(field.value)
+    return temporary_board_values
+
+def get_board_values(board):
+    board_values = []
+    for row in board:
+        for field in row:
+            board_values.append(field.value)
+    return board_values
+
+def validate_players_move(board):
+    temporary_values_board = get_temporary_board_values(board)
+    players_move = str(input("Next move: ").lower())
+    if players_move == "a":
+        move_left(board)
+        sum_left(board)
+        move_left(board)
+        values_board = get_board_values(board)
+        if temporary_values_board == values_board:
```

Comment:
this sequence is repeated 4 times!!! What can you do with it? :D

Reply:
Method was extracted and called check_if_move_is_valid .